File: Cpp/Medium/90. 子集 II.cpp
给定一个可能包含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。

说明：解集不能包含重复的子集。

示例:

输入: [1,2,2]
输出:
[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]

class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        if(nums.empty()) {
            return res;
        }
        sort(nums.begin(), nums.end());
        res.push_back(vector<int>());
        vector<int> v;
        generate(0, v, nums);
        return res;
    }
    void generate(int start, vector<int>& v, vector<int>& nums) {
        if(start == nums.size()) {
            return;
        }
        for(int i = start; i < nums.size(); i++) {
            v.push_back(nums[i]);
            res.push_back(v);
            generate(i + 1, v, nums);
            v.pop_back();
            while(i < nums.size() - 1 && nums[i] == nums[i + 1]) {
                i++;
            }
        }
    }
};

File: Java/medium/90. 子集 II.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        boolean[] isVisited = new boolean[nums.length];
        for(int i = 0; i <= nums.length; i++) {
            dfs(0, temp, res, isVisited, i, nums);
        }
        return res;
    }
    private void dfs(int start, List<Integer> temp, List<List<Integer>> res, boolean[] isVisited, int size, int[] nums) {
        if(temp.size() == size) {
            res.add(new ArrayList<>(temp));
            return;
        }
        for(int i = start; i < nums.length; i++) {
            if(i != 0 && nums[i] == nums[i - 1] && !isVisited[i - 1]) {
                continue;
            }
            temp.add(nums[i]);
            isVisited[i] = true; 
            dfs(i + 1, temp, res, isVisited, size, nums);
            isVisited[i] = false;
            temp.remove(temp.size() - 1);
        }
    }
}
